pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go

```diff
@@ -245,6 +250,22 @@ func applyPriorityFailover(
 	loadAssignment.Endpoints = localityLbEndpoints
 }
 
+// Returning the label names in a separate array as the iteration of map is not ordered.
+func priorityLabelOverrides(labels []string) ([]string, map[string]string) {
+	priorityLabels := make([]string, 0, len(labels))
+	overriddenValueByLabel := make(map[string]string, len(labels))
+	var tempStrings []string
+	for _, labelWithValue := range labels {
+		tempStrings = strings.Split(labelWithValue, string(FailoverPriorityLabelDefaultSeparator))
```

Comment:
Kind of understand this solution. But overall i think it is a little abstract for end-usets,

This API `FailoverPriority []string` can represent two semantics.

How about making `FailoverPriority` oneof like below, which is more concrete and does not break any one either.

```

// same as current FailoverPriority
FailOverPriorityLabels struct {
       Labels []string
}

// same as current FailoverPriority
FailOverPrioritySelectors struct {
       // ordered list of priority from 0 to N
       Selectors []*v1.LabelSelector
}


```

Reply:
IMO its best to keep it in one field to reduce API churn. The API is already complex, this is a niche feature that is ok to hide within the existing field. With proper documentation I think we can make it more clear than 2 fields

Reply:
@hzxuzhonghu 

current FailoverPriority takes labels that define scopes. Priority is assigned to the endpoints as per the intersection of those scopes.
> Endpoints matching all N labels with the client proxy have priority P(0) i.e. the highest priority.
> Endpoints matching the first N-1 labels with the client proxy have priority P(1) i.e. second highest priority.

If addition of another API is acceptable, can the new API `FailOverPrioritySelectors` be added without the scoping/intersection semantic. So the new API is agnostic clientProxy attribute unlike the existing one which is evaluated in the context of the proxy.

Simply endpoints matching the zeroth selector gets priority0 and first selector get priority1 and so on.

Comments?

Reply:
>If addition of another API is acceptable, can the new API FailOverPrioritySelectors be added without the scoping/intersection semantic. So the new API is agnostic of clientProxy attribute unlike the existing one which is evaluated in the context of the proxy.

I am ok with this. @howardjohn @ramaraochavali WDYT

Reply:
> Simply endpoints matching the zeroth selector gets priority0 and first selector get priority1 and so on.

So if there are no endpoints that match a particular slot in the PrioritySelectors - would that priority be ignored? i.e. no end points match FailOverPrioritySelectors[1] , Priority 1 is ignored or it would be incremental priority skipping FailOverPrioritySelectors[1] altogether?

Any case, it seems very niche feature and it seems to OK to have that overridden value after "=" in the existing `failoverPriority` instead of adding a completely new API

Reply:
> Any case, it seems very niche feature and it seems to OK to have that overridden value after "=" in the existing failoverPriority instead of adding a completely new API

In our usecase, in some scenarios, clients in a region need to prefer endpoints in a particular remote region and then local if required. 
Example:  region-1 need to have endpoints in region-2 as priority-0 and region-1 as priority-1. towards the new API. 

Config this with new API is going to be simple. Hence our inclination. 


> So if there are no endpoints that match a particular slot in the PrioritySelectors - would that priority be ignored? i.e. no end points match FailOverPrioritySelectors[1] , Priority 1 is ignored or it would be incremental priority skipping FailOverPrioritySelectors[1] altogether?
These scenarios can be handled similar to current failoverPriority. [Link](https://github.com/istio/istio/blob/34d15dd69a72011da875c2906b0273d3bb6ad734/pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go#L180)

Reply:
> In our usecase, in some scenarios, clients in a region need to prefer endpoints in a particular remote region and then local if required.
Example: region-1 need to have endpoints in region-2 as priority-0 and region-1 as priority-1. towards the new API.

If you need such complicated config, it seems API is better. 

@howardjohn  what do you think?

Reply:
Generally I am pretty concerned about an endless trickle of  slight variations to "failover". There was also a doc presented to the WG of another similar-but-slightly-different failover API

Reply:
We feel the current failoverPriority is tricky and opinionated for 2 reasons:
-  the relation b/w the labels
> Endpoints matching the first N-1 labels with the client proxy have priority P(1) i.e. second highest priority. 
- considering the value of client for selecting endoints.

The new API `FailoverPrioritySelector` is without the above 2 aspects. It is pretty generic and could cover many variations of failover. This can cover the ask in the [doc mentioned by John](https://docs.google.com/document/d/1joq6qDKoebjS8SRD2_u2SvonnArJziPwPuJtepfddaw/edit#heading=h.xw1gqgyqs5b) also.

@howardjohn 
Any specific challanges/concerns you foresee with this new API?

Reply:
So is the proposal just to approve https://docs.google.com/document/d/1joq6qDKoebjS8SRD2_u2SvonnArJziPwPuJtepfddaw/edit#heading=h.j26u6evb6kb1? If so, lets make that proposal approved instead of discussing in the margin of this PR. There was a bit of push back to this in the WG meetings.

Reply:
@hzxuzhonghu 
Addition of the new FailoverPrioritySelector in the context of this use-case is discussed in the istio combined workgroup meet. [Link to discussion](https://youtu.be/jZVqExmnwsM?t=1620)

Members felt this as valid but one off scenario and too soon to introduce a new API. Also suggesting to introduce a separate the failover related APIs in a separate CR and move it out from DestinationRule API.

With that being current stance, can this enhancement to the current failoverPriority be taken forward?

Reply:
https://github.com/istio/istio/pull/43535#discussion_r1116415659 is my main concern. This means we are changing existing API semantics.